node/node.go

```diff
@@ -795,19 +796,38 @@ func (node *Node) getStateTree(ctx context.Context, ts types.TipSet) (state.Tree
 
 // getWeight is the default GetWeight function for the mining worker.
 func (node *Node) getWeight(ctx context.Context, ts types.TipSet) (uint64, error) {
+	h, err := ts.Height()
+	if err != nil {
+		return 0, err
+	}
+	var wFun func(context.Context, types.TipSet, cid.Cid) (uint64, error)
+	v, err := node.VersionTable.VersionAt(types.NewBlockHeight(h))
+	if err != nil {
+		return 0, err
+	}
+	if v >= version.Protocol1 {
+		wFun = node.Chain.ChainSelector.NewWeight
+	} else {
+		wFun = node.Chain.ChainSelector.Weight
+	}
+
 	parent, err := ts.Parents()
 	if err != nil {
 		return uint64(0), err
 	}
 	// TODO handle genesis cid more gracefully
 	if parent.Len() == 0 {
-		return node.Chain.ChainSelector.Weight(ctx, ts, nil)
+		return wFun(ctx, ts, cid.Undef)
 	}
 	pSt, err := node.Chain.ChainReader.GetTipSetState(ctx, parent)
 	if err != nil {
 		return uint64(0), err
 	}
-	return node.Chain.ChainSelector.Weight(ctx, ts, pSt)
+	root, err := pSt.Flush(ctx)
```

Comment:
ChainReader has a `GitTipSetStateRoot` function that avoids the need to instantiate a tree and then flush it. It would also eliminate the need to special case genesis.